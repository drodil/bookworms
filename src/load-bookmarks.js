import got from 'got';
import yaml from 'js-yaml';
import { readFileSync } from 'fs';

const fetchBookmarkConfig = async (path) => {
    if (shouldFetchFromLocal(path)) {
        return {
            type: 'local',
            body: fetchLocaleBookmarkConfig(path)
        }
    } else {
        return {
            type: 'remote',
            body: await fetchRemoteBoomarkConfig(path)
        }
    }
}

const shouldFetchFromLocal = (path) => {
    try {
        new URL (path);
        return false;   
    } catch (e) {
        return true;
    }
}

const fetchLocaleBookmarkConfig = (path) => {
    try {
        return returnResponseAsObject(readFileSync(path, 'utf8'));
    } catch(error) {
        // todo add tests around this error handling
        console.error(`${path} returned ${error.message}`)
        process.exit(1)
    }
}

const fetchRemoteBoomarkConfig = async (path) =>{
    try {
        const { body, statusCode } = await got(path);
        return returnResponseAsJsonOn2xx(body, statusCode, path);
    } catch (error) {
        // todo add tests around this error handling
        console.error(`${path} returned ${error.message}`)
        process.exit(1)
    }
}

const returnResponseAsJsonOn2xx = (body, statusCode, path) => {
    if (statusCode >= 200 && statusCode < 300) {
        return returnResponseAsObject(body, path)
    } else {
        throw new Error(`${path} returned ${statusCode}`)
    }
}

const returnResponseAsObject = (body, path) => {
    if (typeof body === 'object') {
        return body
    } else {
        const json = yaml.load(body);
        // this is bit gross but its what the lib returns
        if (json === "undefined") {
            try {
                return JSON.parse(response);
            } catch(e) {
                throw new Error(`${path} returned invalid bookworms bookmarks structure`)
            }
        } else {
            return json
        }
    }
}

export {fetchBookmarkConfig, shouldFetchFromLocal, fetchLocaleBookmarkConfig, fetchRemoteBoomarkConfig, returnResponseAsJsonOn2xx, returnResponseAsObject}
