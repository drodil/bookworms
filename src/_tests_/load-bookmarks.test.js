import {fetchBookmarkConfig, shouldFetchFromLocal, fetchLocaleBookmarkConfig, fetchRemoteBoomarkConfig,returnResponseAsJsonOn2xx, returnResponseAsObject} from '../load-bookmarks';

describe('loading bookmark config', () => {
    describe('fetchBookmarkConfig', () => {
        test('should return locale bookmarks', async () => {
            expect(await fetchBookmarkConfig('./demo/config/bookmarks.yaml')).toMatchSnapshot();
        })
        test('should return remote bookmarks', async () => {
            expect(await fetchBookmarkConfig('https://raw.githubusercontent.com/thearegee/bookworms/main/demo/config/bookmarks.yaml')).toMatchSnapshot();
        })
    })
    describe('shouldFetchFromLocal', () => {
        test('should return true if path is local filesystem', () => {
            expect(shouldFetchFromLocal('./demo/config/bookmarks.yaml')).toBeTruthy();
        })
        test('should return false if path is remote http', () => {
            expect(shouldFetchFromLocal('https://raw.githubusercontent.com/thearegee/bookworms/main/demo/config/bookmarks.yaml')).toBeFalsy();
        })
    })
    describe('fetchLocaleBookmarkConfig', () => {
        test('should return JSON from local YMAL', () => {
            expect(fetchLocaleBookmarkConfig('./demo/config/bookmarks.yaml')).toMatchSnapshot()
        })
    })
    describe('fetchRemoteBoomarkConfig', () => {
        test('should return JSON from http request', async () => {
            expect(await fetchRemoteBoomarkConfig('https://raw.githubusercontent.com/thearegee/bookworms/main/demo/config/bookmarks.yaml')).toMatchSnapshot()
        })
    })
    describe('returnResponseAsJsonOn2xx', () => {
        test('should throw with statusCode 100', () => {
            try {
                returnResponseAsJsonOn2xx("{'hello':'world'}", 100, 'https://raw.githubusercontent.com/thearegee/bookworms/main/demo/config/bookmarks.yaml')
                expect(true).toBe(false);
            } catch (error) {
                expect(error.message).toEqual('https://raw.githubusercontent.com/thearegee/bookworms/main/demo/config/bookmarks.yaml returned 100')
            }
        })
        test('should throw with statusCode 300', () => {
            try {
                returnResponseAsJsonOn2xx("{'hello':'world'}", 300, 'https://raw.githubusercontent.com/thearegee/bookworms/main/demo/config/bookmarks.yaml')
                expect(true).toBe(false);
            } catch (error) {
                expect(error.message).toEqual('https://raw.githubusercontent.com/thearegee/bookworms/main/demo/config/bookmarks.yaml returned 300')
            }
        })
        test('should return object with string of JSON', () => {
            try {
                const response = returnResponseAsJsonOn2xx("{'hello':'world'}", 200, 'https://raw.githubusercontent.com/thearegee/bookworms/main/demo/config/bookmarks.yaml')
                expect(response).toEqual({hello:"world"})
            } catch (error) {
                expect(true).toBe(false);
            }
        })
    })
    describe('returnResponseAsObject', () => {
        test('should return throw with invalid response error', () => {
            try {
                returnResponseAsObject(undefined, 'https://raw.githubusercontent.com/thearegee/bookworms/main/demo/config/bookmarks.yaml');
                expect(true).toBe(false);
            } catch (error) {
                expect(error.message).toEqual('https://raw.githubusercontent.com/thearegee/bookworms/main/demo/config/bookmarks.yaml returned invalid bookworms bookmarks structure')
            }
        })
        test('should return object with string of JSON', () => {
            try {
                const response = returnResponseAsObject("{'hello':'world'}", 'https://raw.githubusercontent.com/thearegee/bookworms/main/demo/config/bookmarks.yaml');
                expect(response).toEqual({hello:"world"})
            } catch (error) {
                expect(true).toBe(false);
            }
        })
        test('should return object with object', () => {
            try {
                const response = returnResponseAsObject({hello:"world"}, 'https://raw.githubusercontent.com/thearegee/bookworms/main/demo/config/bookmarks.yaml');
                expect(response).toEqual({hello:"world"})
            } catch (error) {
                expect(true).toBe(false);
            }
        })
        test('should return object with YAML', () => {
            try {
                const response = returnResponseAsObject(`hello: world`, 'https://raw.githubusercontent.com/thearegee/bookworms/main/demo/config/bookmarks.yaml');
                expect(response).toEqual({hello:"world"})
            } catch (error) {
                expect(true).toBe(false);
            }
        })
    })
})