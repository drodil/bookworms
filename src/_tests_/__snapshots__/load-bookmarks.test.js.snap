// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loading bookmark config fetchBookmarkConfig should return locale bookmarks 1`] = `
Object {
  "body": Object {
    "description": "These are sample bookmarks to teach you how bookmark works",
    "folders": Array [
      Object {
        "description": "This is to describe the folder structure",
        "folders": Array [
          Object {
            "bookmarks": Array [
              Object {
                "description": "this is used to describe the bookmark",
                "href": "https://www.mywebsite.com",
                "label": "sample url 1",
              },
            ],
            "description": "This is to describe the sub folder structure",
            "label": "sub folder 1",
          },
        ],
        "label": "folder 1",
      },
      Object {
        "folders": Array [
          Object {
            "bookmarks": Array [
              Object {
                "description": "this is used to describe the bookmark",
                "href": "https://www.mysecondwebsite.com",
                "label": "sample url 2",
              },
              Object {
                "description": "this is used to describe the bookmark",
                "href": "https://www.mythirdwebsite.com",
                "label": "sample url 3",
              },
            ],
            "label": "sub folder 2",
          },
          Object {
            "bookmarks": Array [
              Object {
                "description": "this is used to describe the bookmark",
                "href": "https://www.mysecondwebsite.com",
                "label": "sample url 4",
              },
              Object {
                "description": "this is used to describe the bookmark",
                "href": "https://www.mythirdwebsite.com",
                "label": "sample url 5",
              },
            ],
            "label": "sub folder 3",
          },
        ],
        "label": "folder 2",
      },
    ],
    "label": "Bookworms",
  },
  "type": "local",
}
`;

exports[`loading bookmark config fetchBookmarkConfig should return remote bookmarks 1`] = `
Object {
  "body": Object {
    "description": "These are sample bookmarks to teach you how bookmark works",
    "folders": Array [
      Object {
        "description": "This is to describe the folder structure",
        "folders": Array [
          Object {
            "bookmarks": Array [
              Object {
                "description": "this is used to describe the bookmark",
                "href": "https://www.mywebsite.com",
                "label": "sample url 1",
              },
            ],
            "description": "This is to describe the sub folder structure",
            "label": "sub folder 1",
          },
        ],
        "label": "folder 1",
      },
      Object {
        "folders": Array [
          Object {
            "bookmarks": Array [
              Object {
                "description": "this is used to describe the bookmark",
                "href": "https://www.mysecondwebsite.com",
                "label": "sample url 2",
              },
              Object {
                "description": "this is used to describe the bookmark",
                "href": "https://www.mythirdwebsite.com",
                "label": "sample url 3",
              },
            ],
            "label": "sub folder 2",
          },
          Object {
            "bookmarks": Array [
              Object {
                "description": "this is used to describe the bookmark",
                "href": "https://www.mysecondwebsite.com",
                "label": "sample url 4",
              },
              Object {
                "description": "this is used to describe the bookmark",
                "href": "https://www.mythirdwebsite.com",
                "label": "sample url 5",
              },
            ],
            "label": "sub folder 3",
          },
        ],
        "label": "folder 2",
      },
    ],
    "label": "Bookworms",
  },
  "type": "remote",
}
`;

exports[`loading bookmark config fetchLocaleBookmarkConfig should return JSON from local YMAL 1`] = `
Object {
  "description": "These are sample bookmarks to teach you how bookmark works",
  "folders": Array [
    Object {
      "description": "This is to describe the folder structure",
      "folders": Array [
        Object {
          "bookmarks": Array [
            Object {
              "description": "this is used to describe the bookmark",
              "href": "https://www.mywebsite.com",
              "label": "sample url 1",
            },
          ],
          "description": "This is to describe the sub folder structure",
          "label": "sub folder 1",
        },
      ],
      "label": "folder 1",
    },
    Object {
      "folders": Array [
        Object {
          "bookmarks": Array [
            Object {
              "description": "this is used to describe the bookmark",
              "href": "https://www.mysecondwebsite.com",
              "label": "sample url 2",
            },
            Object {
              "description": "this is used to describe the bookmark",
              "href": "https://www.mythirdwebsite.com",
              "label": "sample url 3",
            },
          ],
          "label": "sub folder 2",
        },
        Object {
          "bookmarks": Array [
            Object {
              "description": "this is used to describe the bookmark",
              "href": "https://www.mysecondwebsite.com",
              "label": "sample url 4",
            },
            Object {
              "description": "this is used to describe the bookmark",
              "href": "https://www.mythirdwebsite.com",
              "label": "sample url 5",
            },
          ],
          "label": "sub folder 3",
        },
      ],
      "label": "folder 2",
    },
  ],
  "label": "Bookworms",
}
`;

exports[`loading bookmark config fetchRemoteBoomarkConfig should return JSON from http request 1`] = `
Object {
  "description": "These are sample bookmarks to teach you how bookmark works",
  "folders": Array [
    Object {
      "description": "This is to describe the folder structure",
      "folders": Array [
        Object {
          "bookmarks": Array [
            Object {
              "description": "this is used to describe the bookmark",
              "href": "https://www.mywebsite.com",
              "label": "sample url 1",
            },
          ],
          "description": "This is to describe the sub folder structure",
          "label": "sub folder 1",
        },
      ],
      "label": "folder 1",
    },
    Object {
      "folders": Array [
        Object {
          "bookmarks": Array [
            Object {
              "description": "this is used to describe the bookmark",
              "href": "https://www.mysecondwebsite.com",
              "label": "sample url 2",
            },
            Object {
              "description": "this is used to describe the bookmark",
              "href": "https://www.mythirdwebsite.com",
              "label": "sample url 3",
            },
          ],
          "label": "sub folder 2",
        },
        Object {
          "bookmarks": Array [
            Object {
              "description": "this is used to describe the bookmark",
              "href": "https://www.mysecondwebsite.com",
              "label": "sample url 4",
            },
            Object {
              "description": "this is used to describe the bookmark",
              "href": "https://www.mythirdwebsite.com",
              "label": "sample url 5",
            },
          ],
          "label": "sub folder 3",
        },
      ],
      "label": "folder 2",
    },
  ],
  "label": "Bookworms",
}
`;
